Add support for transfers restricted to mesh divisions

GiudGiud opened the pull request:
closes #25901 (capability needed can be done using the EEIDs for the division)

Pretty big expansion of the capabilities of transfers here:
- new algorithm for finding source processes: not looking at BBOX with a heuristic but matching subapp index and division index
- ability to spatially restrict source and target in completely arbitrary ways. Much more flexible than boundary + block restrictions
- ability to match regions between numbering schemes, which can also be arbitrary. This is not super super useful for shape evaluation (we match things through the location, and a numbering scheme does not change that), but makes nearest-location more useful. There can still be a gain for shape evaluation in that we can restrict the number of potential source subapps (better heuristic than boundinx boxes)
- ability to match source/target regions to a particular target/source subapp. This is going to be helpful to match assemblies to individual subapps. 

Follow up tasks:
- remove nearest-positions transfer and use mesh_divisions for that. That also means re-forwarding nearest_app to yet again a new way
- expand _group_subapps in NLTransfer to all options

GiudGiud committed:
Add support for restricting transfers to inside mesh divisions
refs #25883

GiudGiud committed:
Keep track of whether the mesh is fully indexed
The concern is for higher order variables being transferred from a division that would cut across an element
refs #25883

GiudGiud committed:
Add tests for a variety of uses of mesh divisions in transfers
refs #25883

GiudGiud committed:
Add outgoing data about which point we came from
refs #25883

GiudGiud committed:
Send the origin point mesh division index in source problem
Compare with index in target mesh and only communicate if same: is that what we want???
Some fixes for coordinate transforms and NN
refs #25883

GiudGiud committed:
Factor out acceptance criteria for nearest-position like transfer
refs #25883

GiudGiud committed:
Add different source and target mesh division behavior
- just spatial restriction. Possible for both source and target, in any configuration
- matching mesh division to app index, possible when having multiple source or target
- matching source and target mesh division index, possible in any configuration (one source, multiple source, one tartget multiple target)
refs #25883

GiudGiud committed:
Add extensive parameter checking on the source/target mesh division behavior
refs #25883

GiudGiud committed:
Consolidate meshDivision checks
refs #25883

GiudGiud committed:
Make mesh division parameter clearer
Add better comments and explain more in app source selection in transfer
refs #25883

GiudGiud committed:
Add more options for source app determination algorithm
- bbox is existing option
- greedy search more explicitly greedy
(only in debug mode? it would loose too much scalability otherwise?)
- source mesh divisions, various options
refs #25883

GiudGiud committed:
Add support for source mesh division restrictions in nearest-location field transfer
refs #25883

GiudGiud committed:
Fix typos, add docstring to conservative transfers

GiudGiud committed:
Add utility routine for global to local
This should almost never be used
Use new utilities in legacy code

GiudGiud committed:
Add more options for the algorithm to select which processes should send data
refs #25883

GiudGiud committed:
Add regression tests for using mesh_divisions + nearest-location
refs #25883

GiudGiud committed:
Add exception checking to mesh divisions + transfers
refs #25883

GiudGiud committed:
Fixup exception checks
Add additional behavior on inMeshDivision to handle having passed a subapp/division index for the target and comparing it in the source
refs #25883

GiudGiud committed:
Add regression tests for mesh division + shape evaluation transfer
refs #25883

GiudGiud committed:
Add doc for using mesh divisions
refs #25883

GiudGiud committed:
Use -1 for the invalid division index when outputting
refs #25883

GiudGiud committed:
Correctly use initial positions for the number of positions
refs #25883

GiudGiud committed:
Fix support for nearest location transfer + mesh divisions:
the problem was that we want to match mesh divisions index, but we lose track of where the data stored in the KDTree came from
We could only look at the target point and see if it was in the right source division. But if that target point was outside the source mesh,
just like in many uses of the nearest-location transfer, then we dont get a useful index for the source division. So we HAVE TO use the source mesh
division on the data we feed into the KDTree.

Now the problem is that if we only have N_division KDTree (a parallel to have N_apps KDTree in the normal case) we must mix data from multiple apps
in the KDTree. This is not really what a user would expect imo, especially in a reactor case  where data from different assemblies should not be blended.

So I had to make N_div * N_app KDTree. This was a pretty large rework. The code got real long so now there's a bunch more functions to make it more readable
refs #25883

GiudGiud committed:
Remove a case we likely will never use for shape-evaluation
Shape evaluation are local, we dont need to match indexes
refs #25883

GiudGiud committed:
Fix logic in algorithm for matching subapp index in to_mesh_div case
Add more comments
Check for unsupported situation (boundary restrict, taget mesh div) + high_order earlier
Expand spatial restriction of mesh_divisions to all behaviors
refs #25883

GiudGiud committed:
Add test for matching subapp index & divisions for shape evaluation transfers

GiudGiud committed:
Fix subapp matching to mesh-div for shape evaluation & UO transfers
Add more docstrings
Use less routines since from/to symmetry cannot be done on acceptance criteria

Issues fixed:
-forgot to transform the point to origin frame
-mixed logic to handle both directions wasnt able to handle the to_mesh_div = subapp_index case

Format change on num positions

Add case with only to_mesh_divisons matching divison_index

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1919009/) on 387533a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26221/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26221/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 8cad10d94553be088dd177ac8edd3833a542c3b4`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1919105/) on 89bf186 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/26221/site)

This comment will be updated on new commits.

GiudGiud committed:
More clean up to pass tests:
- use legacy app-grouping mode for nearest-positions tests
- be more careful with the index of the position before looking at whether it is the closest
- add more context to error message in Positions
- ban nearest positions + mesh_divisions in nearest-locations as it s not tested
- revert trying to pre-allocate the buffers. There s something I dont understand, it causes diffs

GiudGiud committed:
For loop formatting and more constness

GiudGiud committed:
Parameter group for mesh divisions parameter
Use to_mesh_division restriction to matching_div_index even if the from_mesh_division is set to restriction
Correctly use the global index for the mesh transform

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1929568/) on 3aae312 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>d450a5</th>
      <th colspan=4>#26221 3aae31</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.01%</td>
      <td>85.02%</td>
      <td>+0.01%</td>
      <td>91.49%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>97252</td>
      <td>97461</td>
      <td>+209</td>
      <td>355</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>17153</td>
      <td>17175</td>
      <td>+22</td>
      <td>33</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/26221/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/26221/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/26221/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented:
nooo, 2 more failures still

GiudGiud committed:
Add explanation for how two arrays are indexed
Add a few more const qualifiers
Rework algorithm matching target divisions to subapps to use the correct global app index
and to get the point division in the source frame

GiudGiud committed:
Disallow test at high parallel count for now

lindsayad commented on framework/include/transfers/MultiAppGeneralFieldNearestLocationTransfer.h:
```diff
@@ -55,8 +56,33 @@ class MultiAppGeneralFieldNearestLocationTransfer : public MultiAppGeneralFieldT
+      const std::vector<std::pair<Point, unsigned int>> & incoming_points,
+      std::vector<std::pair<Real, Real>> & outgoing_vals);
+
+  /// Number of KDTrees used to hold the locations & variable value data
```
```suggestion
  /// Number of KDTrees used to hold the locations and variable value data
```

lindsayad commented on framework/include/transfers/MultiAppGeneralFieldTransfer.h:
```diff
@@ -182,12 +226,36 @@ class MultiAppGeneralFieldTransfer : public MultiAppConservativeTransfer
+  std::vector<const MeshDivision *> _to_mesh_divisions;
+
+  /// How to use the origin mesh divisions to restrict the transfer
+  const MooseEnum & _from_mesh_division_behavior;
+
+  /// How to use the target mesh divisions to restrict the transfer
+  const MooseEnum & _to_mesh_division_behavior;
+
+  /// Matching enum for the mesh division behaviors.
+  enum MeshDivisionTransferUse
+  {
+    RESTRICTION,
+    MATCH_DIVISION_INDEX,
+    MATCH_SUBAPP_INDEX
+  };
```
@dschwen when are we getting your magic enum?

lindsayad reviewed (commented):
Not done

GiudGiud commented on framework/include/positions/Positions.h:
```diff
@@ -33,7 +33,10 @@ class Positions : public GeneralReporter
-  unsigned int getNumPositions() const { return _positions.size(); }
+  unsigned int getNumPositions(bool initial = false) const
+  {
+    return (initial && _initial_positions) ? _initial_positions->size() : _positions.size();
```
which is the reason behind this initial position field. You start multiapps with a positions file, then let the multiapp positions keep track of the real positions

lindsayad replied:
Yes I would just assert that whatever ends up being used is valid

GiudGiud replied:
so I went back and forth on this but "asserting that positions" is valid means checking that initialize() was called on the positions object.

So it s basically a object initialization ordering check, that we do not do for other systems

I already guard every call to _initial_positions->something() with a _initial_positions.
For _positions I cannot do better than manually keeping track of a _initialized boolean

GiudGiud replied:
done

GiudGiud committed:
Address Alex's review

- better doco format

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented on framework/include/transfers/MultiAppGeneralFieldTransfer.h:
```diff
@@ -272,11 +341,11 @@ class MultiAppGeneralFieldTransfer : public MultiAppConservativeTransfer
   std::vector<Real> _fixed_bbox_size;
 
-  /// Number of froms per processor
+  /// Number of froms per processor. Outer indexing is processor number, inner is local app
```
there is one index, that is sometyhing like processor_number * n_local_apps + i_local_app

this is what I mean by innner and outer

lindsayad replied:
Someone just reading this doxygen won't know that. Probably should keep the old doxygen string here and just comment in the `.C` or add what you just said here to the doxygen

GiudGiud replied:
I ll add it to the doxygen

GiudGiud replied:
actually this was the wrong array. Anyway fixed now

GiudGiud committed:
Make positions system 'safer' by manually keeping track of initialization
- this should only be used in debug mode, to check that things were initialized as expected

GiudGiud commented on framework/include/transfers/MultiAppGeneralFieldTransfer.h:
```diff
@@ -225,14 +293,15 @@ class MultiAppGeneralFieldTransfer : public MultiAppConservativeTransfer
     unsigned int problem_id;   // problem id
     dof_id_type dof_object_id; // node or elem id
-    dof_id_type offset;        // Useful when there are more than one point in a given dof object
+    dof_id_type offset; // Useful when there are more than one point in a given dof object (unused)
```
this will tentatively be useful for multiple dofs per point

GiudGiud replied:
sure

GiudGiud replied:
done

GiudGiud committed:
Address Alex's review:
- docstring @param
- define >1 liner in source file
- remove unused point info attribute
- fix docstring explanation on indexing

Update framework/include/transfers/MultiAppGeneralFieldTransfer.h

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

Update framework/include/transfers/MultiAppGeneralFieldTransfer.h

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

GiudGiud commented on framework/include/transfers/MultiAppTransfer.h:
```diff
@@ -195,13 +200,25 @@ class MultiAppTransfer : public Transfer
+    return _current_direction == TO_MULTIAPP
+               ? 0
+               : _from_local2global_map[i_from] - _from_local2global_map[0];
+  }
```
done

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldNearestLocationTransfer.C:
```diff
@@ -436,3 +513,108 @@ MultiAppGeneralFieldNearestLocationTransfer::inBlocks(const std::set<SubdomainID
+{
+
+  if (!_nearest_positions_obj &&
+      (!_from_transforms[getGlobalSourceAppIndex(i_from)]->hasCoordinateSystemTypeChange() ||
+       _skip_coordinate_collapsing))
+    return _from_transforms[getGlobalSourceAppIndex(i_from)]->mapBack(pt);
```
So if there is a coordinate system type change then the transform will never be used? Is there an option for the user to change that behavior if they wanted to? Also it seems that if `_skip_coordinate_collapsing = true`, then the transform object will never be used even if there are no coordinate system type changes. It's possible I'm reading the condition wrong

GiudGiud replied:
if _skip_coordinate_collapsing == true, then the first if-condition is true and we do:
```
    return _from_transforms[getGlobalSourceAppIndex(i_from)]->mapBack(pt);
```
so it should be used. The from transformed will have been set to skip the coordinate collapsing that will skip still.

mapBack errors on going from RZ to XYZ, so this precludes using it. I agree beta/alpha/gamma rotation + scaling is being ignored in that case. We could add more APIs to coordinate transforms to probe more what's in the transform and trigger only the parts we want

GiudGiud replied:
if we had an API like "do the transform map back except if it will error then dont do that erroring part but do everything else" I would use that.

lindsayad replied:
👍

GiudGiud commented on framework/src/positions/Positions.C:
```diff
@@ -17,8 +17,8 @@ Positions::validParams()
 
   // leverage reporter interface to keep track of consumers
-  params.addParam<ReporterName>("initial_positions",
-                                "Positions at the beginning of the simulation");
+  params.addParam<PositionsName>("initial_positions",
+                                 "Positions at the beginning of the simulation");
```
I ll move it

GiudGiud commented on framework/src/positions/Positions.C:
```diff
@@ -40,19 +40,22 @@ Positions::validParams()
 Positions::Positions(const InputParameters & parameters)
   : GeneralReporter(parameters),
     _initial_positions(isParamValid("initial_positions")
```
```suggestion
    // leverage reporter interface to keep track of consumers
    _initial_positions(isParamValid("initial_positions")
```

GiudGiud commented on framework/src/transfers/MultiAppGeneralFieldNearestLocationTransfer.C:
```diff
@@ -52,20 +53,43 @@ MultiAppGeneralFieldNearestLocationTransfer::validParams()
                "We do not support using both nearest positions matching and checking if target "
                "points are within an app domain because the KDTrees for nearest-positions matching "
                "are (currently) built with data from multiple applications.");
+  if (_nearest_positions_obj &&
+      (isParamValid("from_mesh_divisions") || isParamValid("to_mesh_divisions")))
+    paramError("use_nearest_position", "Cannot use nearest positions with mesh divisions");
```
Both are completementary

but for the transfers we only need MeshDivisions, we can do everything "nearest-Positions" using MeshDivisions

GiudGiud commented on framework/src/transfers/MultiAppGeneralFieldNearestLocationTransfer.C:
```diff
@@ -173,35 +205,33 @@ MultiAppGeneralFieldNearestLocationTransfer::prepareEvaluationOfInterpValues(
-      // we could continue here when the positions is not the nearest to the app
+      // Get the current app index
+      const auto i_from = _group_subapps ? app_i : (i_source / getNumDivisions());
+      // Current position index, if using nearest positions
+      const auto i_pos = _group_subapps ? i_source : (i_source % getNumDivisions());
```
it does. This i_pos should not be used elsewhere

GiudGiud commented on framework/src/transfers/MultiAppGeneralFieldNearestLocationTransfer.C:
```diff
@@ -251,11 +304,32 @@ MultiAppGeneralFieldNearestLocationTransfer::buildKDTrees(const unsigned int var
+            const auto tree_division_index = i_source % getNumDivisions();
+
+            // Spatial restriction is always active
+            if (_from_mesh_divisions[i_from]->divisionIndex(*elem) ==
+                MooseMeshDivision::INVALID_DIVISION_INDEX)
+              continue;
+            // We fill one tree per division index for matching subapp index or division index. We
+            // only accept source data from the division index
+            else if ((_from_mesh_division_behavior ==
+                          MeshDivisionTransferUse::MATCH_DIVISION_INDEX ||
+                      _from_mesh_division_behavior ==
+                          MeshDivisionTransferUse::MATCH_SUBAPP_INDEX) &&
+                     tree_division_index != _from_mesh_divisions[i_from]->divisionIndex(*elem))
```
implementation - dependent but hopefully not.

GiudGiud replied:
calling them less now

GiudGiud commented on framework/src/transfers/MultiAppGeneralFieldNearestLocationTransfer.C:
```diff
@@ -251,11 +304,32 @@ MultiAppGeneralFieldNearestLocationTransfer::buildKDTrees(const unsigned int var
+          if (!_from_mesh_divisions.empty())
+          {
+            const auto tree_division_index = i_source % getNumDivisions();
+
```
```suggestion
          const auto elem_div_index = _from_mesh_divisions[i_from]->divisionIndex(*elem);

```

GiudGiud committed:
Address Alex's review

- compute the mesh division for an object less often
- fix comments
- only gather bounding boxes and starting arrays once per variable
- use ternary for computing mesh division in auxkernel
- more dots in sentences

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1955170/) on 5e4fabe wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26221/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26221/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7d1e620dc3e25b783ffb8251ccb6fae15127be4f`

GiudGiud commented on framework/src/auxkernels/MeshDivisionAux.C:
```diff
@@ -35,8 +39,13 @@ MeshDivisionAux::MeshDivisionAux(const InputParameters & parameters)
 {
+  unsigned int value;
   if (isNodal())
-    return _mesh_division.divisionIndex(*_current_node);
+    value = _mesh_division.divisionIndex(*_current_node);
   else
-    return _mesh_division.divisionIndex(*_current_elem);
+    value = _mesh_division.divisionIndex(*_current_elem);
```
done

GiudGiud commented on framework/src/meshdivisions/CartesianGridDivision.C:
```diff
@@ -107,6 +107,18 @@ CartesianGridDivision::initialize()
+    {
+      if (_center_positions)
+        _mesh_fully_indexed = false;
+      else if (_bottom_left(i) > _mesh.getInflatedProcessorBoundingBox(0).first(i) ||
+               _top_right(i) < _mesh.getInflatedProcessorBoundingBox(0).second(i))
+        _mesh_fully_indexed = false;
```
done

GiudGiud commented on framework/src/meshdivisions/NestedDivision.C:
```diff
@@ -45,6 +45,12 @@ NestedDivision::initialize()
   }
   setNumDivisions(tot_divs);
+
+  // If any division does not cover the entire mesh, nested will these same holes
```
changed

GiudGiud commented on framework/src/meshdivisions/NestedDivision.C:
```diff
@@ -45,6 +45,12 @@ NestedDivision::initialize()
+  // If any division does not cover the entire mesh, nested will these same holes
+  _mesh_fully_indexed = true;
+  for (const auto division : _divisions)
+    if (!division->coversEntireMesh())
+      _mesh_fully_indexed = false;
```
done

lindsayad commented on framework/src/transfers/MultiAppGeneralFieldUserObjectTransfer.C:
```diff
@@ -102,43 +103,26 @@ MultiAppGeneralFieldUserObjectTransfer::evaluateInterpValuesWithUserObjects(
          (!point_found || _search_value_conflicts || _nearest_positions_obj);
          ++i_from)
     {
+      Real distance = 1;
```
Why 1?

GiudGiud replied:
it s a fine number. I can do 0

GiudGiud replied:
looks like 0 breaks conflict detection

lindsayad replied:
So this doesn't impact the outcome of `acceptPointInOriginMesh`?

GiudGiud replied:
should not. acceptPointInOriginMesh should provide the distance when using nearest-position. It should not use the distance

lindsayad commented on test/tests/transfers/general_field/nearest_node/mesh_division/main_match_division.i:
```diff
@@ -0,0 +1,168 @@
+
+# Explaining results on the main app:
+# Each value on the main app comes from a region with the same division index on a subapp
+# Because the subapp is not very discretized in Y, some source mesh divisions index
```
```suggestion
# Because the subapp is not very discretized in Y, some source mesh division indices
```

lindsayad commented on test/tests/transfers/general_field/nearest_node/mesh_division/main_match_division.i:
```diff
@@ -0,0 +1,168 @@
+  dim = 2
+  nx = 10
+  ny = 10
+  allow_renumbering = false
```
Is this necessary?

GiudGiud replied:
nope. It helps csvdiffs done with samplers, but exodiff is fine

GiudGiud replied:
resolving because I removed it

lindsayad commented on test/tests/transfers/general_field/nearest_node/mesh_division/tests:
```diff
@@ -0,0 +1,163 @@
+      expect_err = "Attempting to match source subapp index with the number of target mesh "
+                   "divisions, which is 4 while there are 1 source subapp"
+      detail = "the number of bins in the target mesh division and the number of source subapps to "
+               "match are not equal "
```
```suggestion
               "match are not equal."
```

lindsayad commented on test/tests/transfers/general_field/nearest_node/mesh_division/tests:
```diff
@@ -0,0 +1,163 @@
+                 "Transfers/to_sub/to_mesh_division=middle_sub "
+      expect_err = "We do not support using target subapp index for source division behavior and "
+                   "matching the division index"
+      detail = 'the techniques for the target and source mesh divisions are incompatible,'
```
```suggestion
      detail = 'the techniques for the target and source mesh divisions are incompatible.'
```

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1955218/) on 6fa1f5a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26221/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26221/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7d1e620dc3e25b783ffb8251ccb6fae15127be4f`

GiudGiud commented on framework/src/transfers/MultiAppGeneralFieldTransfer.C:
```diff
@@ -430,20 +624,37 @@ MultiAppGeneralFieldTransfer::cacheOutgoingPointInfo(const Point point,
-    // Store point information
+    // Select which from_mesh_division the source data must come from for this point
+    unsigned int required_source_division = 0;
+    if (_from_mesh_division_behavior == MeshDivisionTransferUse::MATCH_SUBAPP_INDEX)
+      required_source_division = getGlobalTargetAppIndex(problem_id);
+    else if (_from_mesh_division_behavior == MeshDivisionTransferUse::MATCH_DIVISION_INDEX ||
+             _to_mesh_division_behavior == MeshDivisionTransferUse::MATCH_DIVISION_INDEX ||
+             _to_mesh_division_behavior == MeshDivisionTransferUse::MATCH_SUBAPP_INDEX)
+      required_source_division = _to_mesh_divisions[problem_id]->divisionIndex(
+          _to_transforms[getGlobalTargetAppIndex(problem_id)]->mapBack(point));
```
yes there's the option to just use spatial restriction

GiudGiud replied:
in that case, that we make it with required_source_divison = 0 because we are just using it as a spatial restriction, then `required_source_divison` will not be used later on so it does not matter its value

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1955418/) on ce802de wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/26221/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/26221/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7d1e620dc3e25b783ffb8251ccb6fae15127be4f`

GiudGiud commented on framework/src/transfers/MultiAppGeneralFieldTransfer.C:
```diff
@@ -1195,25 +1437,56 @@ MultiAppGeneralFieldTransfer::setSolutionVectorValues(
+
+    // Get nearest position (often a subapp position) for the target point
+    // We want values from the child app that is closest to the same position as the target
+    // TODO: use transformed point
+    Point nearest_position_source;
+    if (_nearest_positions_obj)
+    {
+      const bool initial = _fe_problem.getCurrentExecuteOnFlag() == EXEC_INITIAL;
+      const Point nearest_position = _nearest_positions_obj->getNearestPosition(pt, initial);
+      nearest_position_source = _nearest_positions_obj->getNearestPosition(
+          (*_from_transforms[from_global_num])(Point(0, 0, 0)), initial);
```
nice

GiudGiud commented on framework/src/transfers/MultiAppGeneralFieldTransfer.C:
```diff
@@ -1195,25 +1437,56 @@ MultiAppGeneralFieldTransfer::setSolutionVectorValues(
+      nearest_position_source = _nearest_positions_obj->getNearestPosition(
+          (*_from_transforms[from_global_num])(Point(0, 0, 0)), initial);
+
+      if (!_skip_coordinate_collapsing &&
+          _from_transforms[from_global_num]->hasNonTranslationTransformation())
+        mooseError("Rotation and scaling currently unsupported with nearest positions transfer.");
```
no. Nearest_position is an option to the GeneralTransfers. Hopefully we ll get rid of it and nest it under the support for mesh divisions instead

GiudGiud replied:
(for which we should be supporting all transformations, except the ill-determined mapBack)

GiudGiud commented on framework/src/transfers/MultiAppGeneralFieldTransfer.C:
```diff
@@ -313,20 +445,23 @@ MultiAppGeneralFieldTransfer::transferVariable(unsigned int i)
   _froms_per_proc.clear();
   _froms_per_proc = getFromsPerProc();
 
+  // Get the index for the first source app every processor owns
+  _global_app_start_per_proc = getGlobalStartAppPerProc();
```
ah good point. Actually the bounding boxes could as well. I moved this out

lindsayad replied:
It's a bummer that more can't be done in `initialSetup`. You've made it as good as it can be given when `getAppInfo` is called

GiudGiud commented on framework/src/transfers/MultiAppGeneralFieldNearestLocationTransfer.C:
```diff
@@ -173,35 +205,33 @@ MultiAppGeneralFieldNearestLocationTransfer::prepareEvaluationOfInterpValues(
-  for (unsigned int i_source = 0; i_source < num_sources; ++i_source)
+  // Construct a local KDTree for each source. A source can be a single app or multiple apps
+  // combined (option for nearest-position / mesh-divisions)
+  for (const auto i_source : make_range(num_sources))
```
the whole file uses them now!

GiudGiud commented:
thank you for the reviews!

loganharbour commented:
Broke next https://civet.inl.gov/event/159269/

You guys really need to add sweeps by default to these :D

GiudGiud commented:
I had added sweeps I did think distributed would be different. These transfers should behave roughly the same. Partitioning must be different??

lindsayad commented:
partitioning can definitely be different. Metis vs. ParMetis
